File: scripts/mapConvertion.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import Header
from nav_msgs.msg import OccupancyGrid
from octomap_msgs.msg import Octomap
import octomap

def octomap_callback(msg):
    rospy.loginfo("Received octomap message, processing...")

    octree = octomap.OcTree(msg.data)
    
    # Define OccupancyGrid
    occupancy_grid = OccupancyGrid()
    occupancy_grid.header = Header(stamp=rospy.Time.now(), frame_id='map')
    occupancy_grid.info.resolution = msg.resolution

    # Convert Octree to OccupancyGrid
    for leaf in octree.iter_leafs():
        x, y, z = leaf.get_coordinate()
        if z < z_threshold: # Consider the floor and objects
            i = int((x - min_x) / msg.resolution)
            j = int((y - min_y) / msg.resolution)
            occupancy_grid.data[j * width + i] = 100 if leaf.get_value() > occupancy_threshold else 0

    # Publish the occupancy grid
    occupancy_grid_pub.publish(occupancy_grid)

if __name__ == "__main__":
    rospy.init_node('octomap_to_gridmap')

    # Define parameters
    min_x = rospy.get_param("~min_x", -10)
    min_y = rospy.get_param("~min_y", -10)
    width = rospy.get_param("~width", 20)
    height = rospy.get_param("~height", 20)
    z_threshold = rospy.get_param("~z_threshold", 0.5)  # Ignore everything below this threshold
    occupancy_threshold = rospy.get_param("~occupancy_threshold", 0.5)

    # Initialize the grid map
    occupancy_grid = OccupancyGrid()
    occupancy_grid.info.width = width
    occupancy_grid.info.height = height
    occupancy_grid.info.resolution = 0.05  # Default value, will be overwritten
    occupancy_grid.data = [0 for _ in range(width * height)]  # Initialize with free space

    # Subscribe and publish
    rospy.Subscriber('/octomap_full', Octomap, octomap_callback)
    occupancy_grid_pub = rospy.Publisher('/occupancy_grid', OccupancyGrid, queue_size=10)

    rospy.spin()

File: src/mapConvertion.cpp
#include <ros/ros.h>
#include <octomap_msgs/Octomap.h>
#include <nav_msgs/OccupancyGrid.h>
#include <iostream>
#include <vector>
#include <std_msgs/Float64.h>

ros::Publisher map_pub;
nav_msgs::OccupancyGrid map2d;

void octomapCallback(const octomap_msgs::Octomap::ConstPtr& msg) {
    try {
        // Convert the Octomap to a vector
        std::vector<int8_t> octomap_data(msg->data.begin(), msg->data.end());

        // Extract the dimensions from the Octomap message
        int num_cells = octomap_data.size();

        // Assuming the Octomap is a full occupancy octree, set the binary flag accordingly
        bool binary = true;

        // Extract the resolution from the Octomap message
        double resolution = msg->resolution;

        int width = std::sqrt(num_cells);
        int height = num_cells / width;

        std::vector<int8_t> map2d_data(width * height, 0);
        for (int i = 0; i < num_cells; ++i) {
            map2d_data[i] = octomap_data[i];
        }

        // Normalize the values to range from 0 to 100
        int max_value = *std::max_element(map2d_data.begin(), map2d_data.end());
        for (int i = 0; i < num_cells; ++i) {
            map2d_data[i] = static_cast<int8_t>((map2d_data[i] / static_cast<double>(max_value)) * 100);
        }

        // Update the 2D map message
        map2d.header = msg->header;
        map2d.info.resolution = resolution;
        map2d.info.width = width;
        map2d.info.height = height;
        map2d.data = map2d_data;

        // Publish the 2D map
        map_pub.publish(map2d);
    }
    catch (const std::exception& e) {
        ROS_ERROR_STREAM("Failed to convert Octomap to 2D map: " << e.what());
    }
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "octomap_to_2dmap");
    ros::NodeHandle nh;

    // Create publisher for 2D map
    map_pub = nh.advertise<nav_msgs::OccupancyGrid>("/map", 10);

    // Create subscriber to octomap_full topic
    ros::Subscriber sub = nh.subscribe("/octomap_full", 10, octomapCallback);

    ros::spin();

    return 0;
}